src/tools/rbd_nbd/rbd-nbd.cc

```diff
@@ -792,34 +794,106 @@ static int parse_imgpath(const std::string &imgpath)
   return 0;
 }
 
-static int do_list_mapped_devices()
+static void get_mapped_info(int pid, std::string &pool_name, std::string &image_name, std::string &snap_name)
 {
-  char path[64];
-  int m = 0;
-  int fd[2];
+  std::string path = "/proc/" + stringify(pid) + "/cmdline";
+  std::ifstream ifs;
+  std::string cmdline;
 
-  if (socketpair(AF_UNIX, SOCK_STREAM, 0, fd) == -1) {
-    int r = -errno;
-    cerr << "rbd-nbd: socketpair failed: " << cpp_strerror(-r) << std::endl;
-    return r;
+  ifs.open(path.c_str(), std::ifstream::in);
+  assert (ifs.is_open());
+  ifs >> cmdline;
+
+  for(unsigned int index = 0; index < cmdline.size(); index++) {
+    if(((unsigned char)cmdline[index]) < ' ') {
+      cmdline[index] = ' ';
```

Comment:
I would suggest to create `vector<const char*> args`, and fill it with pointers to command line parameters.
Then you can reuse the code from `rbd_nbd()` to parse command line arguments. That is you will need extract that code to a separate function.

For parsing image spec you can reuse `parse_imgpath()` function (after some modification). And actually, for me it would be ok if we leave the image spec unparsed, as it is provided by a user, but I don't have a strong opinion here.